src/App/StringHasher.h

```diff
@@ -0,0 +1,489 @@
+
+    QVector<StringIDRef> relatedIDs() const {
+        if (_sid)
+            return _sid->relatedIDs();
+        return QVector<StringIDRef>();
+    }
+
+    bool isBinary() const {
+        if (_sid)
+            return _sid->isBinary();
+        return false;
+    }
+
+    bool isHashed() const {
+        if (_sid)
+            return _sid->isHashed();
+        return false;
+    }
+
+    void toBytes(QByteArray &bytes) const {
+        if (_sid)
+            _sid->toBytes(bytes, _index);
+    }
+
+    PyObject *getPyObject(void) {
+        if (_sid)
+            return _sid->getPyObjectWithIndex(_index);
+        Py_INCREF(Py_None);
+        return Py_None;
+    }
+
+    void mark() const {
+        if (_sid)
+            _sid->mark();
+    }
+
+    bool isMarked() const {
+        return _sid && _sid->isMarked();
+    }
+
+    bool isFromSameHasher(const StringHasherRef & hasher) const
+    {
+        return _sid && _sid->isFromSameHasher(hasher);
+    }
+
+    StringHasherRef getHasher() const
+    {
+        if (_sid)
+            return _sid->getHasher();
+        return StringHasherRef();
+    }
+
+    void setPersistent(bool enable)
+    {
+        if (_sid)
+            _sid->setPersistent(enable);
+    }
+
+    friend class StringHasher;
+
+private:
+    StringID *_sid;
+    int _index;
+};
+
+/// A String table to map string from/to a unique integer
+class AppExport StringHasher: public Base::Persistence, public Base::Handled {
+
+    TYPESYSTEM_HEADER_WITH_OVERRIDE();
+
+public:
+    StringHasher();
+    virtual ~StringHasher();
+
+    virtual unsigned int getMemSize (void) const override;
+    virtual void Save (Base::Writer &/*writer*/) const override;
+    virtual void Restore(Base::XMLReader &/*reader*/) override;
+    virtual void SaveDocFile (Base::Writer &/*writer*/) const override;
+    virtual void RestoreDocFile (Base::Reader &/*reader*/) override;
+    void setPersistenceFileName(const char *name) const;
+    const std::string &getPersistenceFileName() const;
+
+    /** Maps an arbitrary string to an integer
+     *
+     * The function maps an arbitrary text string to a unique integer ID, which
+     * is returned as a shared pointer to reference count the ID so that it is
+     * possible to prune any unused strings.
+     *
+     * If the string is longer than the threshold setting of this StringHasher,
+     * it will be sha1 hashed before storing, and the original content of the
+     * string is discarded.
+     *
+     * The purpose of function is to provide a short form of a stable string
+     * identification.
+     */
+    StringIDRef getID(const char *text, int len=-1, bool hashable=false);
+
+    /** Map text or binary data to an integer */
+    StringIDRef getID(const QByteArray & data, bool binary, bool hashable=true, bool nocopy=false);
+
+    /** Map geometry element name to an integer */
+    StringIDRef getID(const Data::MappedName & name,
+                      const QVector<StringIDRef> & sids);
```

Comment:
@realthunder can you add documentation of the parameters here? Also, @wwmayer, are you concerned about boolean blindness in the first two of these functions? The first one only has the single boolean at the end, but the second has three, which I suspect in client code is pretty opaque.